### PR TITLE
Ticket1770 block server restart crash

### PR DESCRIPTION
**Pull request by AdrianPotter**
### Description of work

The call to restart an IOC returns immediately. When we check waitfor_running, we may query the IOC before it's actually stopped as part of that restart. To avoid this situation, we check the read back on the restart IOC PV to see if there are any operations pending.

In the course of the investigation, I've also added a bit more error handling code.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1770

### Acceptance criteria

- [x] Blockserver can be restarted consistently without it crashing (this happened for me about 25% of the time when restarting from the console pre-fix).
- [x] No changes to the availability of blocks from the GUI
- [x] Confirm that after restart that essential IOCs are still running (RUNCTRL, BLOCKSVR, BLOCKCACHE, ISISDAE)

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Has the author taken into account the multi-thredded nature of the code?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [x] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by FreddieAkeroyd** [comment]
The code currently uses the PV  DEVIOS:SysReset as a check, however this can exist and look the same both before and after a reset and hence the reset is missed, which i think underlies the issue. There are other PVs that may be more apapropriate to use. One for example might be DEVIOS:STARTTOD which is the date/time the IOC was started, so when this changes the IOC must have finished restarting.

File: BlockServer/epics/procserv_utils.py [hunk 1]
```diff
@@ -64,6 +64,18 @@ def restart_ioc(self, prefix, ioc):
...
+        """Tests to see if an IOC restart is pending
+
+        Args:
+            prefix (string): The prefix for the instrument
+            ioc (string): The name of the IOC
+
+        Returns:
+            bool: Whether a restart is pending
+        """
+        True if ChannelAccess.caget(self.generate_prefix(prefix, ioc) + ":RESTART", as_string=True) is "Busy" else False
```
Comment (DominicOram) [inline BlockServer/epics/procserv_utils.py]: Is there a missing return statement here?

File: BlockServer/runcontrol/runcontrol_manager.py [hunk 2]
```diff
@@ -221,11 +222,13 @@ def wait_for_ioc_start(self):
         while True:
             # See if the IOC has restarted by looking for a standard PV
             try:
-                ans = self._channel_access.caget(self._prefix + RC_PV)
+                running = True if self._channel_access.caget(self._prefix + RC_RESET_PV) is not None else False
+                restart_pending = True if self._channel_access.caget(self._prefix + RC_RESTART_PV) is "Busy" else False
```
Comment (DominicOram) [inline BlockServer/runcontrol/runcontrol_manager.py]: This seems very similar to the code in procserv_utils, can we reuse?

**Comment by DominicOram** [comment]
Couldn't reliably reproduce original error. Moving back into rework due to Freddie's comment and a couple of comments I had on the code

File: BlockServer/epics/procserv_utils.py (hunk 1 above)
Comment (AdrianPotter) [inline BlockServer/epics/procserv_utils.py]: Yes, that would be helpful wouldn't it

File: BlockServer/runcontrol/runcontrol_manager.py (hunk 2 above)
Comment (AdrianPotter) [inline BlockServer/runcontrol/runcontrol_manager.py]: I'll have a look

**Comment by AdrianPotter** [comment]
In response to @FreddieAkeroyd's comment, we could potentially look at the IOC start date. I've opted to look at :RESTART because it's the actual PV that's used to initiate the IOC restart so seems like the best option for knowing if a restart is pending. Assuming that to be the case, we only need a proc serv PV to tell us whether the IOC is active, which we already have.

I don't think using the IOC start date exclusively is practical. We'd have to preserve the initial value from before the restart is requested in order to know when the restart is finished. That means a thread can't check if a restart is pending unless it was involved in the original operation and so has the previous start time value. Although the alternative is monitoring 2 PVs, I think the added overhead is worth the extra flexibility. I'm open to suggestions though.

**Comment by AdrianPotter** [comment]
Also @DominicOram, how are you trying to reproduce the issue? My workflow was:

1. Start IBEX
2. Access the block server with `console -M localhost BLOCKSVR`
3. Restart with `ctrl+X`

I could see the blockserver initiating an additional restart around 25% of the time.

**Comment by FreddieAkeroyd** [comment]
@AdrianPotter yes, looking at :RESTART will tell you if a restart is pending – I thought I had already changed the code to wait for a restart to complete (via use of a completion callback) as part of   ISISComputingGroup/IBEX#1710   but it seems the pull request for this  ISISComputingGroup/EPICS-inst_servers#87   wasn’t merged when the ticket was reviewed and closed. 

We just need to be caution about introducing a potential race condition. When the procServControl restart PV goes back to DONE it just means ^X has been sent to procServ, not that anything has actually happened yet. When procServ has done the restart it issues a message to stdout which is then picked up by procServControl to adjust the IOC status PV. Thus if you called get_ioc_status() too soon after RESTART is DONE I think you may incorrectly get the pre-restart RUNNING status, and may even get a pre-restart read of the SysReset PV. Putting in a small delay between these operations may solve the issue. The advantage I saw of something like STARTTOD was that if you read the incorrect version you know and can discard it and continue waiting. I don’t think a thread needs to be involved in the original operation to check the start time – it just requires the thread that requests the original restart writing the “time of restart request” into an IOC name keyed dictionary (protect by appropriate locks)  and any thread that wishes to see if the ioc is still restarting just need to check that   STARTTOD PV  >=  restart_time_dict{‘ioc_name’}  to know when it has finished.

**Pull request merged**